[20.09] Fix Anndata index access, use numpy-safe JSON encoder

mtekman opened the pull request:
Re-added try/except block due to failing datasets: https://usegalaxy.eu/u/pmoreno/h/test-older-and-newer-anndata-from-zero

references issue here https://github.com/galaxyproject/galaxy/issues/10639

mtekman committed:
Re-added try/except block due to failing datasets

mtekman committed:
blank lines linter fix

pcm32 commented on lib/galaxy/datatypes/binary.py:
```diff
@@ -981,78 +981,81 @@ def sniff(self, filename):
+
+                if 'obs' in dataset.metadata.layers_names:
+                    tmp = anndata_file["obs"]
+                    dataset.metadata.obs_names = [util.unicodify(x) for x in tmp["index"]]
```
```suggestion
                    tmp = anndata_file["obs"]
                    if "index" in tmp:
                        obs_index = "index"
                    else:
                        obs_index = "_index"
                    dataset.metadata.obs_names = [util.unicodify(x) for x in tmp[obs_index]]
```
Maybe we can take the chance to cater for the "_index" 0.7.4 case

pcm32 replied:
feel free to neglect as per below comments.

mtekman committed:
Update lib/galaxy/datatypes/binary.py

Co-authored-by: Pablo Moreno <pcm32@users.noreply.github.com>

mtekman committed:
index test for variables added too

mtekman committed:
added new anndata test data for 0.6, 0.7.4, and unknown

      These are valid AnnData test datasets that look normal when
      opened with the AnnData python library

mtekman committed:
unindent, remove try/catch, test indexes before further processing in
obs and var

mtekman committed:
linting, readability

mvdbeek committed:
Use customtypes.json_encoder to encode metadata dict

This is what we use internally to store values in the database, and this
deals with bytestrings, numpy integers etc., and fixes the new Anndata
tests (which pass if you use `retry_metadata_internally`).

mvdbeek committed:
Drop unicodify workaround that shouldn't be necessart after 9254b05dd0e0d97a26a7ce3aa5ba1ed3ae2c6652

nsoranzo commented on lib/galaxy/model/metadata.py:
```diff
@@ -212,9 +211,9 @@ def to_JSON_dict(self, filename=None):
         if '__validated_state_message__' in dataset_meta_dict:
             meta_dict['__validated_state_message__'] = dataset_meta_dict['__validated_state_message__']
         if filename is None:
```
```suggestion
        encoded_meta_dict = galaxy.model.custom_types.json_encoder.encode(meta_dict)
        if filename is None:
```

nsoranzo commented on lib/galaxy/model/metadata.py:
```diff
@@ -212,9 +211,9 @@ def to_JSON_dict(self, filename=None):
             meta_dict['__validated_state_message__'] = dataset_meta_dict['__validated_state_message__']
         if filename is None:
-            return json.dumps(meta_dict)
+            return galaxy.model.custom_types.json_encoder.encode(meta_dict)
```
```suggestion
            return encoded_meta_dict
```

nsoranzo commented on lib/galaxy/model/metadata.py:
```diff
@@ -212,9 +211,9 @@ def to_JSON_dict(self, filename=None):
+            return galaxy.model.custom_types.json_encoder.encode(meta_dict)
         with open(filename, 'wt+') as fh:
-            json.dump(meta_dict, fh)
+            fh.write(galaxy.model.custom_types.json_encoder.encode(meta_dict))
```
```suggestion
            fh.write(encoded_meta_dict)
```

nsoranzo commented on lib/galaxy/datatypes/binary.py:
```diff
@@ -998,21 +1010,29 @@ def set_meta(self, dataset, overwrite=True, **kwd):
-                dataset.metadata.obs_layers, \
-                    dataset.metadata.obs_count, \
-                    dataset.metadata.obs_size = _layercountsize(tmp, len(dataset.metadata.obs_names))
+                obs_index = False
```
```suggestion
                obs_index = None
```

nsoranzo commented on lib/galaxy/datatypes/binary.py:
```diff
@@ -1022,17 +1042,25 @@ def _layercountsize(tmp, lennames=0):
-                dataset.metadata.var_layers, \
-                    dataset.metadata.var_count, \
-                    dataset.metadata.var_size = _layercountsize(tmp, len(tmp["index"]))
+                var_index = False
```
```suggestion
                var_index = None
```

mtekman committed:
Update lib/galaxy/datatypes/binary.py

Co-authored-by: Nicola Soranzo <nicola.soranzo@gmail.com>

mtekman committed:
Update lib/galaxy/datatypes/binary.py

Co-authored-by: Nicola Soranzo <nicola.soranzo@gmail.com>

mvdbeek committed:
Simplify encoded_metadata_dict handling

Co-authored-by: Nicola Soranzo <nicola.soranzo@gmail.com>

mvdbeek committed:
Also use numpy-safe encoder in model store

mvdbeek reviewed (approved):
Thank you @mtekman & @pcm32 & @nsoranzo !
